### ISSUE TITLE
How to properly integrate with frontend javascript

### ISSUE THREAD

**Issue by phuang-hive** [kanidm/webauthn-rs#468]
**TLDR:
I am getting the `WebauthnError::AuthenticationFailure` error and I can't figure out how to fix it. Using this crate in my rust backend server with a frontend javascript that does the browser `Webauthn` calls.**

Full Context:

Hello! I've been using trying to integrate this crate (rust backend) with a javascript frontend. I've managed to get passkey registration working (with `start_passkey_registration` and `finish_passkey_registration`). However, I haven't been able to get an authentication flow to work yet (`start_passkey_authentication` and `finish_passkey_authentication`).

When I call `finish_passkey_authentication` I am getting the `WebauthnError::AuthenticationFailure` error. I looked through the github / found the place where this error is happen (I think [here](https://github.com/kanidm/webauthn-rs/blob/96ddb68104e62dc2aa855da070cda3057cc44095/webauthn-rs-core/src/core.rs#L934-L935)). This seems to be because the signature and the hash of the `clientDataJSON` object are not matching.

For both `start_passkey_registration` and `start_passkey_authentication`, the documentation [here](https://docs.rs/webauthn-rs/latest/webauthn_rs/struct.Webauthn.html#method.start_passkey_authentication) and [here](https://docs.rs/webauthn-rs/latest/webauthn_rs/struct.Webauthn.html#method.start_passkey_registration) both say we can simply serialize the `CreationChallengeResponse` and the `RequestChallengeResponse` structs respectively and give it to the browser to do the `navigator.credentials.create()` and `navigator.credentials.get()` calls. In particular, here is a quote: "This function returns a CreationChallengeResponse which you must serialise to json and send to the user agent (e.g. a browser) for it to conduct the registration" 

However, reading from this guide [here](https://webauthn.guide/#webauthn-api) and the official specs [here](https://w3c.github.io/webauthn/#dom-publickeycredentialrequestoptions-challenge), the `challenge` must be a `BufferSource` type. However in my frontend Javascript, when I receive the`CreationChallengeResponse` and the `RequestChallengeResponse` from backend, it is a JSON type, which means the challenge is a `String` which is not a `BufferSource`. The [guide here](https://webauthn.guide/#webauthn-api) gives an example of converting something to a `BufferSource` i.e like this code snippet
`Uint8Array.from(
        CreationChallengeResponse.publicKey.challenge => c.charCodeAt(0))`
        
 Doing these conversations allowed me to call the `navigator.credentials.create() and `navigator.credentials.get()` in my frontend javascript. However, now I needed to transmit back to my rust backend (I use actix-web for the webserver as I think you guys too in the tutorial).
 
The response objects from `navigator.credentials.create()` and `navigator.credentials.get()` both have `ArrayBuffer` values (such as for `rawId`. Currently my backend is receiving the `RegisterPublicKeyCredential` and `PublicKeyCredential` as actix `Json` which you guys seem to [do in the tutorial as well](https://github.com/kanidm/webauthn-rs/blob/96ddb68104e62dc2aa855da070cda3057cc44095/tutorial/server/actix_web/src/handler/auth.rs#L116-L117). However, I can't send `ArrayBuffer` types using the fetch API on frontend Javascript, so I'm doing so extra logic after the webauthn browser API calls to convert the `ArrayBuffer` values into base64 strings that can be sent in a JSON.

One other extra step I have to do is that I was previously getting a `WebauthnError::MismatchedChallenge` error from [here](https://github.com/kanidm/webauthn-rs/blob/96ddb68104e62dc2aa855da070cda3057cc44095/webauthn-rs-core/src/core.rs#L480-L481). The only way I was able to fix it was by doing this somewhat hacky procedure of taking the challenge value from the `clientDataJSON` from the `navigator.credentials.create()` call, converting it to base64 and then recreating the whole `clientDataJSON` again in a new object, before sending it to backend.

These few extra steps work for registration, but it is not working for authentication. I am assuming because of this "recreating clientDataJSON", its somehow changing the bytes of the `clientDataJSON` sent to backend, which is used to do the authentication signature verification as described at the top of the post.

With all this context, I'm wondering what I can do to fix this? The tutorial (unless I'm missing something) doesn't show how to use the crate with a javascript frontend where the browser is calling the `Webauthn` API and transmitting the information backend and forth with a backend using this crate (the tutorial only seems to cover a rust frontend as well?). So I'm wondering if I could get any guidance on how to properly setup my javascript frontend to interact with this crate so that I won't get this authentication WebauthnError::AuthenticationFailure` error.

Appreciate any and all help I can get on this! Please let me know if I should provide more info or context or anything. Thanks!

**Comment by phuang-hive** [comment]
I've tried my best to use the tutorial as a guide but unfortunately, seems like since the "frontend" components are in rust too, it doesn't answer my question of how to properly do it in Javascript and then transmit the results to the backend using HTTP calls. For example [here in the tutorial](https://github.com/kanidm/webauthn-rs/blob/96ddb68104e62dc2aa855da070cda3057cc44095/tutorial/wasm/src/lib.rs#L291-L292), a simple two lines of 
```
let w_rpkc = web_sys::PublicKeyCredential::from(jsval);
                    // Serialise the web_sys::pkc into the webauthn proto version, ready to
                    // handle/transmit.
                    let pkc = PublicKeyCredential::from(w_rpkc);
```
did everything that I'm struggling with :((

**Comment by phuang-hive** [comment]
I keep reading [this guide](https://webauthn.guide/#webauthn-api) over and over again, and I'm convinced I'm getting the authentication error because I am modifying the clientDataJSON

> clientDataJSON: As during registration, the clientDataJSON is a collection of the data passed from the browser to the authenticator. It is one of the items used during authentication as the source bytes to generate the signature.

However I can't not modify the clientDataJSON without then getting the challenge mismatch error?

**Comment by phuang-hive** [comment]
I figured out my mistake, it was that the response we get from `start_passkey_authentication` and `start_passkey_registration`, the challenges are **also** base64. So I had to use the `window.atob()` function on those challenges before passing them into the navigator calls as well.

**Comment by Firstyear** [comment]
Yes, because the webauthn specification is silly and assumed that you could pass in a uint8buffer from json (which is actually literally impossible). So everyone had to write base64 adapters.
